### ISSUE TITLE
VM being disposed early

### ISSUE THREAD

**Issue by OracPrime** [dsuryd/dotNetify#31]
Hi, great library, really enjoying the simplification it provides.

However it's not clear to me what manages the lifetimes of the VMs.  I am instantiating a long-lived calculation engine in my VM, and I need it to stay alive for the duration of a connection from a browser.  However it seems to be getting disposed at apparently random times after a bit of inactivity.

What can make it dispose other than a call to vm.$dispose() [ which I have in componentWillUnmount, but that's not where it's coming from]

Using dotnetify 2.0.4beta with react 15.5.4

Thanks.

**Comment by dsuryd** [comment]
Hi, the default is to auto-dispose the VMs if there's no activity after 20 minutes.  I should have exposed the  settings to override - that'll be my next to-do,  but you can set up a background javascript timer to ping the VM to avoid expiration.

**Comment by OracPrime** [comment]
[realised I replied by email, not here, so duplicating the essence here in case it helps others]
I was seeing VM disposal after much less than 20 minutes - maybe 1 or 2 minutes.

**Comment by dsuryd** [comment]
Sorry, I didn't receive the email you mentioned. If there's more details in there, would you include it here.  It's surprising if it gets disposed only within 1-2 minutes, since it's pretty noticeable.   Do you see this only on your app, or on any of my demo app as well?  If you just start the server, will it happen within that time frame too?

**Comment by OracPrime** [comment]
I only saw it in my app, as I had just browsed your apps online rather than running them.  I'll see if I can get any more specific data for you.

**Comment by OracPrime** [comment]
OK, it's not an idle timeout [caveat - I'm running this from Visual Studio under the debugger, in IIS Express, so this may be affecting things].  There is a signalr ping message from the browser every 5 minutes, which prevents the 20 minute timeout being reached.  Of the form

http://localhost:48352/signalr/ping?_=1499874190068

**Comment by OracPrime** [comment]
Aha!  However, on the first dotnetify action after the 20 minute timeout, *that* is when the VM_Disposed message occurs (which I am writing out in the Dispose method of my VM).  So with the IDE just sitting there, no apparent dispose.  As soon as I attempt to push a new value from the browser, at that point it goes through the dispose method.

**Comment by dsuryd** [comment]
The behavior of the expiration time in the .NET MemoryCache that's used to store the VMs is that it won't dispose right at the expiration time, but only on the first attempt after the fact; so this explains what you saw.  And you are correct that signalR ping won't reset the sliding expiration.

As a workaround, you can add this to the Startup.cs Configure method to set to long expiration time:
```
public void Configure( IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory )
{
   ((VMControllerFactory)app.ApplicationServices.GetService<IVMControllerFactory>()).CacheExpiration = new System.TimeSpan(24,0,0,0);
```
I'll soon add this as a configurable settings with default to never expire.

**Comment by OracPrime** [comment]
Thank you very much!  It's all working very nicely now, thank you.

**Comment by dsuryd** [comment]
Fix included in v2.2.0-pre.
